MapboxCoreNavigationTests/RouteControllerTests.swift

```diff
@@ -113,4 +134,69 @@ class RouteControllerTests: XCTestCase {
+        let tunnelEntranceLocation = CLLocation(latitude: intersectionLocation.latitude, longitude: intersectionLocation.longitude)
+        
+        navigation.locationManager(navigation.locationManager, didUpdateLocations: [tunnelEntranceLocation])
+
+        let currentLegProgress = (navigation.routeProgress.currentLegProgress)!
+        let currentStep = currentLegProgress.currentStep
+        let currentIntersection = currentStep.intersections![currentLegProgress.stepIndex]
+        
+        var currentLocation = location(at: navigation.location!.coordinate)
+        var distanceToTunnelEntrance = currentIntersection.location.distance(to: currentLocation.coordinate)
+        
+        var userIsAtTunnelEntranceRadius = navigation.userWithinTunnelEntranceRadius(at: currentLocation, intersection: currentIntersection, distance: distanceToTunnelEntrance)
+        XCTAssertTrue(userIsAtTunnelEntranceRadius, "Location must be within the tunnel entrance radius")
+
+        let outsideTunnelEntranceRadius =  intersectionLocation.coordinate(at: 200, facing: intersectionLocation.direction(to: tunnelSetup.firstLocation.coordinate))
+        let outsideTunnelEntranceRadiusLocation = CLLocation(latitude: outsideTunnelEntranceRadius.latitude, longitude: outsideTunnelEntranceRadius.longitude)
+        
+        navigation.locationManager(navigation.locationManager, didUpdateLocations: [outsideTunnelEntranceRadiusLocation])
+        
+        currentLocation = location(at: navigation.location!.coordinate)
+        distanceToTunnelEntrance = currentIntersection.location.distance(to: outsideTunnelEntranceRadiusLocation.coordinate)
+
+        userIsAtTunnelEntranceRadius = navigation.userWithinTunnelEntranceRadius(at: currentLocation, intersection: currentIntersection, distance: distanceToTunnelEntrance)
+        XCTAssertFalse(userIsAtTunnelEntranceRadius, "Location must not outside the tunnel entrance radius")
+    }
+    
+    fileprivate func location(at coordinate: CLLocationCoordinate2D) -> CLLocation {
+        return CLLocation(coordinate: coordinate,
+                            altitude: 5,
+                  horizontalAccuracy: 258.20,
+                    verticalAccuracy: 200,
+                              course: 20,
+                               speed: 15,
+                           timestamp: Date())
+    }
+    
+    // TODO: Test Tunnel Animation Enabled
+    // TODO: Test Tunnel Animation Disabled
+    func testTunnelSimulatedNavigation() {
+        
+    }
+    
+    func testEnableTunnelAnimation() {
+        let navigation = tunnelSetup.routeController
+        
+        // Intersection with a tunnel roadClass
+        navigation.advanceStepIndex(to: 1)
+        let intersectionLocation = navigation.routeProgress.currentLegProgress.currentStep.intersections![1].location
+        navigation.routeProgress.currentLegProgress.currentStepProgress.distanceTraveled = intersectionLocation.distance(to: tunnelSetup.firstLocation.coordinate)
```

Comment:
Again, don't mutate things on RouteProgress in tests. Just supply a new location.

Reply:
In this case I'm inclined to agree. So would Demeter.